controllers/nstemplateset/namespaces.go

```diff
@@ -189,31 +191,50 @@ func (r *namespacesManager) ensureInnerNamespaceResources(ctx context.Context, n
 		return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to process template for namespace '%s'", nsName)
 	}
 
-	if currentRef, exists := namespace.Labels[toolchainv1alpha1.TemplateRefLabelKey]; exists && currentRef != "" && currentRef != tierTemplate.templateRef {
-		logger.Info("checking obsolete namespace resources", "spacename", nsTmplSet.GetName(), "tier", nsTmplSet.Spec.TierName, "type", tierTemplate.typeName)
-		if err := r.setStatusUpdatingIfNotProvisioning(ctx, nsTmplSet); err != nil {
-			return err
-		}
-		currentTierTemplate, err := getTierTemplate(ctx, r.GetHostCluster, currentRef)
-		if err != nil {
-			return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusUpdateFailed, err, "failed to retrieve current TierTemplate with name '%s'", currentRef)
-		}
-		currentObjs, err := currentTierTemplate.process(r.Scheme, map[string]string{
-			SpaceName: nsTmplSet.GetName(),
-		}, template.RetainAllButNamespaces)
-		if err != nil {
-			return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusUpdateFailed, err, "failed to process template for TierTemplate with name '%s'", currentRef)
+	// Check if we need to delete any obsolete resources
+	// First check if the namespace already has template reference meaning it's not freshly created and has not got any inner resources applied yet
+	currentRef, exists := namespace.Labels[toolchainv1alpha1.TemplateRefLabelKey]
+	if exists {
+		// Now check if there is any featured object to be deleted in case the feature annotation does not present in the NSTemplateSet anymore
+		toDeleteObsoleteFeaturedObjects := false
+		featureStr := nsTmplSet.Annotations[toolchainv1alpha1.FeatureToggleNameAnnotationKey]
+		features := strings.Split(featureStr, ",")
```

Comment:
should we check that the features annotations is not nil before splitting it ?

Reply:
Not really. If the map is nil then `nsTmplSet.Annotations[toolchainv1alpha1.FeatureToggleNameAnnotationKey]` would return an empty string.

Reply:
You can "read" a nil map. You can't "write" to a nil map.
```
m := nil
v := m["key"] // always returns ""
m["key"] = "value" // panics
```